transports/quic/src/stream_map.rs

```diff
@@ -0,0 +1,522 @@
+            Some(stream) => Either::Left(stream),
+            None => {
+                let (sender, receiver) = oneshot::channel();
+                self.connectors.push_front(sender);
+                self.wake_driver();
+                Either::Right(receiver)
+            }
+        }
+    }
+
+    pub(crate) fn close_reason(&self) -> Result<(), Error> {
+        match self
+            .close_reason
+            .as_ref()
+            .map(|e| Error::ConnectionError(e.clone()))
+        {
+            Some(e) => Err(e),
+            None => Ok(()),
+        }
+    }
+
+    pub(crate) fn destroy_stream(&mut self, stream: StreamId) {
+        trace!("Removing substream {:?} from map", *stream);
+        self.connection.destroy_stream(*stream);
+        self.map.remove(&stream.0).expect(
+            "Internal state corrupted. \
+             You probably used a Substream with the wrong StreamMuxer",
+        );
+    }
+
+    pub(crate) fn close(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+        trace!(
+            "close() called with {} outbound streams for side {:?}",
+            self.connection.send_streams(),
+            self.connection.side()
+        );
+        self.wake_closer();
+        self.close_waker = Some(cx.waker().clone());
+        if self.close_reason.is_some() {
+            return Poll::Ready(Ok(()));
+        } else if self.connection.send_streams() == 0 {
+            self.shutdown(0);
+            self.close_reason = Some(quinn_proto::ConnectionError::LocallyClosed);
+            return Poll::Ready(Ok(()));
+        }
+        for (&stream, value) in &mut self.map {
+            value.wake_all();
+            let _ = self.connection.finish(stream);
+        }
+        Poll::Pending
+    }
+
+    pub(crate) fn handle_event(
+        &mut self,
+        event: quinn_proto::ConnectionEvent,
+    ) -> Option<quinn_proto::EndpointEvent> {
+        self.connection.handle_event(event)
+    }
+
+    pub(crate) fn shutdown_stream(
+        &mut self,
+        cx: &mut Context<'_>,
+        substream: &StreamId,
+    ) -> Result<oneshot::Receiver<()>, quinn_proto::FinishError> {
+        self.wake_driver();
+        self.finish(substream)?;
+        let (sender, mut receiver) = oneshot::channel();
+        let _ = receiver.poll_unpin(cx);
+        self.get(substream).set_finisher(sender);
+        Ok(receiver)
+    }
+}
+
+/// The connection driver
+#[derive(Debug)]
+struct ConnectionDriver {
+    inner: Arc<Mutex<Streams>>,
+    /// The packet awaiting transmission, if any.
+    outgoing_packet: Option<quinn_proto::Transmit>,
+    /// The timer being used by this connection
+    timer: Option<futures_timer::Delay>,
+    /// The last timeout returned by `quinn_proto::poll_timeout`.
+    last_timeout: Option<Instant>,
+    /// Transmit socket
+    socket: Arc<socket::Socket>,
+}
+
+impl Upgrade {
+    pub(crate) fn spawn(
+        connection: Arc<Mutex<Streams>>,
+        _side: quinn_proto::Side,
```

Comment:
If this parameter is not used why is it not removed?

Reply:
Good question :)
```suggestion

```